### PR TITLE
Unwrap an already-compiled callable before handing it back to torch.compile

### PR DESCRIPTION
**Pull request by danielhanchen**
### Symptom

`GPT_OSS_MXFP4_(20B)-Inference.ipynb` and `GPT_OSS_BNB_(20B)-Inference.ipynb` both die at model load on Colab A100/T4 VMs running torch 2.11.0+cu128 with transformers 4.56.2. The error is a bare `AssertionError:` with no message:

```
unsloth/models/_utils.py       in _run_temporary_patches(phase)
unsloth_zoo/.../gpt_oss.py     in patch_GptOssAttention()
unsloth_zoo/.../common.py      in _compile_or_fall_back(*args, **kwargs)
torch/__init__.py              in compile(...)
torch/_dynamo/eval_frame.py    in __call__(self, fn)
AssertionError:
```

Reproduced locally, off Colab, on an identical stack (torch 2.11.0+cu128, transformers 4.56.2, B200), with the exact same frame list:

```
_utils.py:3446 unsloth_compile_transformers -> _run_temporary_patches("post_compile")
gpt_oss.py:2214 apply_rotary_pos_emb = torch_compile(apply_rotary_pos_emb)
common.py:193   return torch.compile(*args, **kwargs)
eval_frame.py:1073 assert not hasattr(compile_wrapper, "get_compiler_config")
```

### Root cause

The assert is torch's own, and it means "the callable you handed me is already a compiled one". `torch.compile` builds its wrapper with `functools.wraps(fn)`, which copies `fn.__dict__`, then checks the copy did not bring `get_compiler_config` along. Dynamo unwraps its own wrappers before that point (`innermost_fn`), so the assert is only reachable through a `functools.wraps` copy of a compiled function that forwards `get_compiler_config` without being a Dynamo wrapper.

That is exactly what `_fall_back_to_eager_on_recompile_limit` returns. It forwards `get_compiler_config` and `_torchdynamo_orig_callable` deliberately, so that "is this compiled?" checks keep answering yes once the eager fallback is installed.

The GPT-OSS chain:

1. `pre_compile` writes `unsloth_compiled_cache/unsloth_compiled_module_gpt_oss.py`, whose `apply_rotary_pos_emb` carries `@torch_compile_with_fallback(fullgraph = True, ...)`, and installs it on `transformers.models.gpt_oss.modeling_gpt_oss`.
2. `post_compile` re-runs `patch_GptOssAttention`, which imports that name and calls `torch_compile(apply_rotary_pos_emb)`.

Up to torch 2.10 `innermost_fn` followed `_torchdynamo_orig_callable` unconditionally, unwrapped the copy to the eager original and recompiled it, so none of this showed. torch 2.11 added an id check:

```python
while (hasattr(unaltered_fn, "_torchdynamo_orig_callable")
       and getattr(unaltered_fn, "_torchdynamo_wrapper_id", None) == id(unaltered_fn)):
```

whose docstring names the case directly: "When functools.wraps copies `_torchdynamo_orig_callable` from a wrapped function, the copied `_torchdynamo_wrapper_id` won't match the outer wrapper's id." So the wrapper is no longer unwrapped, the copied `get_compiler_config` survives into `compile_wrapper.__dict__`, and the assert fires.

`patch_function` has always hand-unwrapped `get_compiler_config` carriers via `__wrapped__` before compiling (`utils.py:1789`). The bare `torch_compile` / `_torch_compile` decorators never got that unwrap, and `_compile_or_fall_back`'s non-fullgraph branch reached `torch.compile` with nothing around it, so the failure was fatal rather than a lost optimisation.

### Fix

`unsloth_zoo/temporary_patches/common.py`

* New `unwrap_already_compiled(function)`: follow `__wrapped__` while the callable advertises `get_compiler_config`. Unlike `patch_function`'s inline version it tolerates a carrier with no `__wrapped__` (an `OptimizedModule`) and terminates on a `__wrapped__` cycle, because the shared funnel sees everything.
* `_compile_or_fall_back` now builds the decorator for either branch and applies it inside one `decorate()` that unwraps first and catches a refusal to decorate, returning the eager function with a one-per-call-site warning.

Only **decoration** is guarded. Compilation still happens on the first call exactly as before, so a setup where compile succeeds keeps its compiled kernel at the same speed; the fallback engages only when torch declines to build a wrapper at all.

`unsloth_zoo/temporary_patches/utils.py`

* `compile_with_eager_fallback` applies the same unwrap before `_raw_torch_compile`. A no-op for every current caller (they all pass module-level plain functions), and it keeps the hazard out of reach if one ever passes a callable that has been through here already.

### Audit of every compile entry point

`_compile_or_fall_back` has no callers outside `common.py` (it is only reached through the two aliases). `torch_compile` and `_torch_compile` cover 15 call sites; the ones that omit `fullgraph`, and so took the unprotected branch, are `gpt_oss.py:2214`, `gpt_oss.py:2451-2452`, `gpt_oss.py:2528`, `gpt_oss.py:2583`, `misc.py:607/610`, `flex_attention/utils.py:77` and `patch_torch_functions.py:131`. Nothing depends on the unprotected behaviour. `patch_torch_functions.py:131` already guards itself with `if not hasattr(qat_class.forward, "get_compiler_config")`, which is the same hazard spotted once and fixed locally. `_raw_torch_compile` keeps its `functools.partial(torch.compile, ...)` form: its one caller applies the fallback itself, and `test_generated_fullgraph_fallback.py` asserts on that shape.

### Tests

New `tests/test_compile_already_compiled_callable.py`, six tests:

* the unwrap returns the eager original, and survives a carrier with no `__wrapped__` and a `__wrapped__` cycle
* `torch_compile` and `_torch_compile` hand **the eager function** to `torch.compile`, observed by recording the argument, so it fails on every torch rather than only the one where the refusal is fatal
* the GPT-OSS `pre_compile` then `post_compile` sequence does not assert (gated on torch >= 2.11, where the assert exists)
* a refused decoration falls back to eager instead of raising
* a working compile is still compiled, which is the control against turning the fix into a silent performance regression

Compatibility, running the new file on both this branch and 566aa2b2:

| torch | 566aa2b2 | this branch |
| --- | --- | --- |
| 2.6.0+cu124 | 4 failed | 5 passed, 1 skipped |
| 2.7.1+cpu | 4 failed | 5 passed, 1 skipped |
| 2.8.0+cu128 | 4 failed | 5 passed, 1 skipped |
| 2.9.1+cu128 | 4 failed | 5 passed, 1 skipped |
| 2.10.0+cpu | 4 failed | 5 passed, 1 skipped |
| 2.11.0+cu128 | 5 failed | 6 passed |
| 2.12.0+cu130 | 5 failed | 6 passed |
| 2.13.0+cpu | 5 failed | 6 passed |

The extra failure and the missing skip from 2.11 onward are the GPT-OSS sequence test, which only runs where the assert exists.

Existing suites, unchanged: `test_generated_fullgraph_fallback`, `test_recompile_limit_fallback`, `test_eager_fallback_latch`, `test_disabled_hook_graph_break`, `test_rmsnorm_recompile_guards`, `test_compile_disable_partial`, `test_rl_replacements_compile_disable`, `test_checkpoint_compile_mix`, `test_compiler_decorated_forward`, `test_unreadable_modeling_source`. 193 passed 1 skipped on torch 2.11, 239 passed on torch 2.9.

### End to end on a real model

torch 2.11.0+cu128, transformers 4.56.2, one B200, unsloth 2026.8.5:

* `unsloth/gpt-oss-20b-unsloth-bnb-4bit` loads and generates ("What is 2+2?" -> 4). Dies at load on 566aa2b2.
* `unsloth/gpt-oss-20b` loads and generates. Dies at load on 566aa2b2.
* 3 LoRA steps on the BNB model, loss 2.5276 -> 2.4472.

### Not verified

* Not run on Colab. Everything above is a local reproduction of the same stack; the Colab VMs were left alone.
* `triton_kernels` is not installed here, so the MXFP4 repo took the 16-bit LoRA path rather than the true MXFP4 kernel path. The patch sequence that fails is the same either way, but the MXFP4 kernels themselves were not exercised.
* No T4. The Colab reports include a G4 VM; only B200 (and CPU-only venvs for the version matrix) were available.
* No throughput measurement. The claim that a working compile stays compiled rests on the returned object still being a `torch.compile` wrapper, asserted in `test_a_working_compile_is_still_compiled`, not on a timed run.
* torch 2.6 and 2.7 were exercised through the new test file only, not through a GPT-OSS load.

### CODE REVIEW

File: unsloth_zoo/temporary_patches/common.py [hunk 1]
```diff
@@ -177,6 +178,57 @@ def _decorator(func: F) -> F:
...
+    reproduces the pre-2.11 outcome exactly rather than degrading to eager.
+    """
+    seen = set()
+    while callable(function) and hasattr(function, "get_compiler_config"):
+        inner = getattr(function, "__wrapped__", None)
+        # No `__wrapped__` (an OptimizedModule, say) or a cycle: leave it be and
+        # let the caller's guard deal with whatever torch makes of it.
+        if not callable(inner) or id(inner) in seen: break
+        seen.add(id(function))
+        function = inner
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/temporary_patches/common.py:227]: Correct, fixed in b1f7ffaf. A bound method forwards attribute lookups to its underlying function, so `__wrapped__` there is the unbound original and following it drops the receiver. torch's own `innermost_fn` stops on a bound method for the same reason ("id(bound_method) != id(wrapper_function), so we won't unwrap through `__func__` and lose the self binding"), so the loop now stops as well when `__self__` is set, and the bound method reaches torch unchanged. On torch <= 2.10 torch discarded the receiver itself regardless, so this is a no-op there; on 2.11+ the bound method now hits torch's own refusal and the fallback returns it eagerly instead of silently handing back a callable that wants self. Regression test: `test_unwrap_already_compiled_keeps_a_bound_method_bound`.

File: unsloth_zoo/temporary_patches/common.py [hunk 2]
```diff
@@ -186,13 +238,40 @@ def _compile_or_fall_back(*args, **kwargs):
...
+                _UNCOMPILED_CALL_SITES.add(label)
+                logger.warning(
+                    f"Unsloth: torch.compile refused to wrap {label}; running "
+                    f"it eagerly. Training and inference are unaffected apart "
+                    f"from speed. ({type(exception).__name__}: {exception})"
+                )
+            return function
     if args and callable(args[0]):
         return decorate(args[0])
     return decorate
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/temporary_patches/common.py]: Correct, fixed in b1f7ffaf. `torch.compile`'s first parameter is named `model`, so `torch_compile(model = fn)` is a legal spelling and the funnel was only looking in `args`. It returned the undecorated `decorate`, and the failure was worse than losing the compile: the caller then called that, so its first argument was treated as the function to compile and handed straight back. Measured on the previous head, `_torch_compile(model = fn)(torch.zeros(2))` returned `tensor([0., 0.])`. A callable `model` is now popped out of the compile kwargs into the positional slot, for both the fullgraph and the non-fullgraph branch. Regression test: `test_the_model_keyword_spelling_compiles_the_function`.

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review

File: unsloth_zoo/temporary_patches/common.py [hunk 3]
```diff
@@ -186,15 +244,49 @@ def _compile_or_fall_back(*args, **kwargs):
...
-    from .utils import torch_compile_with_fallback
-    decorate = torch_compile_with_fallback(**kwargs)
-    if args and callable(args[0]):
-        return decorate(args[0])
+    # `torch.compile`'s first parameter is named `model`, so `torch_compile(model = fn)`
+    # is a third legal spelling. Take it out of the compile kwargs, where it would
+    # collide with the function passed positionally, and treat it as that function.
+    function = args[0] if args and callable(args[0]) else None
+    if function is None and callable(kwargs.get("model")):
+        function = kwargs.pop("model")
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/temporary_patches/common.py]: Correct, fixed in c96812b3. `model` is now popped whenever the key is present rather than only when it is callable, so `torch_compile(model = None, ...)` returns a clean decorator factory and nothing collides with the function `torch.compile` is given positionally. Worth flagging how quiet the failure was: the collision raised inside the guard, which read it as the compiler declining to decorate, so compilation was off with only a one-time warning. Regression test: `test_the_explicit_model_none_decorator_factory_still_compiles`, which fails on the previous commit.

File: unsloth_zoo/temporary_patches/common.py [hunk 4]
```diff
@@ -177,6 +178,63 @@ def _decorator(func: F) -> F:
...
+        # the receiver. torch's own `innermost_fn` stops on a bound method for that
+        # exact reason ("id(bound_method) != id(wrapper_function), so we won't
+        # unwrap through __func__ and lose the self binding").
+        if getattr(function, "__self__", None) is not None: break
+        inner = getattr(function, "__wrapped__", None)
+        # No `__wrapped__` (an OptimizedModule, say) or a cycle: leave it be and
+        # let the caller's guard deal with whatever torch makes of it.
+        if not callable(inner) or id(inner) in seen: break
+        seen.add(id(function))
+        function = inner
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/temporary_patches/common.py:227]: Not taking this one. `_fall_back_to_eager_on_recompile_limit` is the only `functools.wraps` wrapper in the repo that carries `get_compiler_config` (it copies it at temporary_patches/utils.py:1536 on purpose, so "is this compiled?" checks keep answering yes), and it is transparent by construction: it calls the compiled function and latches to the eager original only once the recompile cache is exhausted. There is no semantic wrapper here for the loop to discard, and `patch_function` has unwrapped any `get_compiler_config` carrier through `__wrapped__` the same way for as long as it has existed (utils.py:1794).

Restricting the unwrap to identity metadata would also remove the only case the change exists for. That wrapper's `_torchdynamo_wrapper_id` is a `functools.wraps` copy of the inner compile wrapper's, so it does not equal `id(wrapper)` - which is exactly the mismatch torch 2.11 added the check for, and exactly why the assert became reachable. Anything keyed on that metadata would skip it and leave gpt-oss failing at load.

Worth noting the behaviour is not new either: through torch 2.10 `innermost_fn` followed `_torchdynamo_orig_callable` unconditionally, so torch itself unwrapped these same carriers on every call site.

**Comment by danielhanchen** [comment]
@codex review
